File: cpp/evm.cpp
// C++20: run `make cpp` from the repository root.
#include "intx.hpp"
#include "json.hpp"
#include <cstddef>
#include <cstdint>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <stdexcept>
#include <string>
#include <utility>
#include <vector>

using intx::uint256;
using nlohmann::json;

struct Result {
    bool success;
    std::vector<uint256> stack;
    std::string return_data;
    json logs = json::array();
    json state = json::object();
};

Result evm(const std::vector<uint8_t>& code) {
    size_t pc = 0;
    std::vector<uint256> stack;

    while (pc < code.size()) {
        [[maybe_unused]] const auto opcode = code[pc++];

        // TODO: implement the EVM here!
    }

    return {true, std::move(stack)};
}

std::vector<uint8_t> decode_hex(const std::string& hex) {
    if (hex.size() % 2 != 0) throw std::runtime_error("Invalid bytecode hex");
    std::vector<uint8_t> bytes;
    for (size_t i = 0; i < hex.size(); i += 2) {
        size_t consumed = 0;
        auto byte = std::stoul(hex.substr(i, 2), &consumed, 16);
        if (consumed != 2) throw std::runtime_error("Invalid bytecode hex");
        bytes.push_back(static_cast<uint8_t>(byte));
    }
    return bytes;
}

void print_stack(const std::vector<uint256>& stack) {
    std::cout << '[';
    for (size_t i = 0; i < stack.size(); ++i) {
        if (i) std::cout << ", ";
        std::cout << intx::to_string(stack[i]);
    }
    std::cout << "]\n";
}

int main() {
    const auto test_file = std::filesystem::exists("../evm-pro.json")
        ? "../evm-pro.json" : "../evm.json";
    std::ifstream input(test_file);
    const auto tests = json::parse(input);

    for (size_t i = 0; i < tests.size(); ++i) {
        const auto& test = tests[i];
        std::cout << "Test #" << i + 1 << '/' << tests.size() << ": " << test.at("name").get<std::string>() << '\n';
        const auto& code = test.at("code");
        const auto& expected = test.at("expect");
        // As tests get more complex, pass more inputs to evm.
        const auto result = evm(decode_hex(code.at("bin").get<std::string>()));
        std::vector<uint256> expected_stack;
        const bool check_stack = expected.contains("stack") && !expected.at("stack").is_null();
        if (check_stack) {
            for (const auto& value : expected.at("stack")) {
                expected_stack.push_back(intx::from_string<uint256>(value.get<std::string>()));
            }
        }
        const bool check_success = expected.contains("success") && !expected.at("success").is_null();
        const bool expected_success = check_success ? expected.at("success").get<bool>() : result.success;
        const json outputs = {{"return", result.return_data}, {"logs", result.logs}, {"state", result.state}};
        bool outputs_match = true;
        for (const auto& [field, actual] : outputs.items()) {
            if (expected.contains(field) && !expected.at(field).is_null() && expected.at(field) != actual) {
                std::cout << field << " mismatch: expected " << expected.at(field) << "; got " << actual << '\n';
                outputs_match = false;
            }
        }

        if (!outputs_match || result.success != expected_success || (check_stack && result.stack != expected_stack)) {
            std::cout << std::boolalpha << "Expected success: " << expected_success << "; got: " << result.success << '\n';
            std::cout << "Expected stack: "; print_stack(expected_stack);
            std::cout << "Actual stack: "; print_stack(result.stack);
            std::cout << "Instructions:\n" << (code.at("asm").is_null() ? code.at("bin") : code.at("asm")).get<std::string>() << '\n';
            std::cout << "Hint: " << test.value("hint", "") << '\n';
            std::cout << "Progress: " << i << '/' << tests.size() << '\n';
            return 1;
        }
    }
    std::cout << "Progress: " << tests.size() << '/' << tests.size() << '\n';
}

File: zig/evm.zig
// Zig 0.16: run `make zig` from the repository root.
const std = @import("std");

const Result = struct {
    success: bool,
    stack: std.ArrayList(u256),
    return_data: []const u8 = "",
    logs: std.json.Value,
    state: std.json.Value,
};

fn evm(code: []const u8, allocator: std.mem.Allocator) !Result {
    var pc: usize = 0;
    const stack = try std.ArrayList(u256).initCapacity(allocator, 1024);

    while (pc < code.len) {
        const opcode = code[pc];
        pc += 1;
        _ = opcode;

        // TODO: implement the EVM here!
    }

    return .{
        .success = true,
        .stack = stack,
        .logs = .{ .array = std.json.Array.init(allocator) },
        .state = .{ .object = std.json.ObjectMap{} },
    };
}

const TestCase = struct {
    name: []const u8,
    hint: []const u8 = "",
    code: struct { bin: []const u8, @"asm": ?[]const u8 = null },
    expect: struct {
        success: ?bool = null,
        stack: ?[]const []const u8 = null,
        @"return": ?[]const u8 = null,
        logs: ?std.json.Value = null,
        state: ?std.json.Value = null,
    },
};

fn jsonEqual(a: std.json.Value, b: std.json.Value) bool {
    if (std.meta.activeTag(a) != std.meta.activeTag(b)) return false;
    return switch (a) {
        .null => true,
        .bool => |v| v == b.bool,
        .integer => |v| v == b.integer,
        .float => |v| v == b.float,
        .number_string => |v| std.mem.eql(u8, v, b.number_string),
        .string => |v| std.mem.eql(u8, v, b.string),
        .array => |v| blk: {
            if (v.items.len != b.array.items.len) break :blk false;
            for (v.items, b.array.items) |left, right| {
                if (!jsonEqual(left, right)) break :blk false;
            }
            break :blk true;
        },
        .object => |v| blk: {
            if (v.count() != b.object.count()) break :blk false;
            var entries = v.iterator();
            while (entries.next()) |entry| {
                const other = b.object.get(entry.key_ptr.*) orelse break :blk false;
                if (!jsonEqual(entry.value_ptr.*, other)) break :blk false;
            }
            break :blk true;
        },
    };
}

pub fn main(init: std.process.Init) !void {
    const allocator = init.arena.allocator();
    const cwd = std.Io.Dir.cwd();
    const json = cwd.readFileAlloc(init.io, "../evm-pro.json", allocator, .unlimited) catch |err| switch (err) {
        error.FileNotFound => try cwd.readFileAlloc(init.io, "../evm.json", allocator, .unlimited),
        else => return err,
    };
    const parsed = try std.json.parseFromSlice([]TestCase, allocator, json, .{ .ignore_unknown_fields = true });
    const tests = parsed.value;

    for (tests, 0..) |test_case, index| {
        std.debug.print("Test #{d}/{d}: {s}\n", .{ index + 1, tests.len, test_case.name });
        const code = try allocator.alloc(u8, test_case.code.bin.len / 2);
        _ = try std.fmt.hexToBytes(code, test_case.code.bin);
        // As tests get more complex, pass more inputs to evm.
        const result = try evm(code, allocator);
        const stack_values = test_case.expect.stack orelse &.{};
        const expected_stack = try allocator.alloc(u256, stack_values.len);
        for (stack_values, 0..) |value, i| {
            expected_stack[i] = try std.fmt.parseInt(u256, value, 0);
        }

        var outputs_match = true;
        if (test_case.expect.@"return") |value| {
            if (!std.mem.eql(u8, result.return_data, value)) {
                std.debug.print("return mismatch: expected {s}; got {s}\n", .{ value, result.return_data });
                outputs_match = false;
            }
        }
        if (test_case.expect.logs) |value| {
            if (!jsonEqual(result.logs, value)) {
                std.debug.print("logs mismatch\n", .{});
                outputs_match = false;
            }
        }
        if (test_case.expect.state) |value| {
            if (!jsonEqual(result.state, value)) {
                std.debug.print("state mismatch\n", .{});
                outputs_match = false;
            }
        }
        const success_matches = if (test_case.expect.success) |value| result.success == value else true;
        if (!outputs_match or !success_matches or (test_case.expect.stack != null and !std.mem.eql(u256, result.stack.items, expected_stack))) {
            std.debug.print("Expected success: {any}; got: {}\n", .{ test_case.expect.success, result.success });
            std.debug.print("Expected stack: {any}\nActual stack: {any}\n", .{ expected_stack, result.stack.items });
            std.debug.print("Instructions:\n{s}\n", .{test_case.code.@"asm" orelse test_case.code.bin});
            std.debug.print("Hint: {s}\nProgress: {d}/{d}\n", .{ test_case.hint, index, tests.len });
            std.process.exit(1);
        }
    }
    std.debug.print("Progress: {d}/{d}\n", .{ tests.len, tests.len });
}
